workflows/namespace.go

```diff
@@ -0,0 +1,307 @@
+		// namespace found, and specs don't match,  update it
+		updateResp, err = w.UpdateNamespace(ctx, &cloudservice.UpdateNamespaceRequest{
+			Namespace:       ns.Namespace,
+			Spec:            spec,
+			ResourceVersion: ns.ResourceVersion,
+		})
+		if err != nil {
+			return out, err
+		}
+		namespaceID = ns.Namespace
+		asyncOpID = updateResp.AsyncOperation.Id
+		out.Outcome = ReconcileOutcomeUpdated
+
+	} else {
+		// nothing to change, get the latest namespace and return
+		namespaceID = ns.Namespace
+		out.Outcome = ReconcileOutcomeUnchanged
+		return out, nil
+	}
+
+	if asyncOpID != "" {
+		// wait for the operation to complete
+		_, err = w.WaitForAsyncOperation(ctx, &WaitForAsyncOperationInput{
+			AsyncOperationID: asyncOpID,
+			Timeout:          namespaceUpdateTimeout,
+		})
+		if err != nil {
+			return out, err
+		}
+	}
+	var getResp *cloudservice.GetNamespaceResponse
+	getResp, err = w.GetNamespace(ctx, &cloudservice.GetNamespaceRequest{
+		Namespace: namespaceID,
+	})
+	if err != nil {
+		return out, err
+	}
+	ns = getResp.Namespace
+	return out, nil
+}
+
+// Reconcile a namespace, create the namespace if one does not exist, or update the namespace if one does exist.
+func (w *workflows) ReconcileNamespace(ctx workflow.Context, in *ReconcileNamespaceInput) (*ReconcileNamespaceOutput, error) {
+	if err := validator.ValidateStruct(in); err != nil {
+		return nil, fmt.Errorf("invalid input: %s", err)
+	}
+	namespace, err := w.GetNamespaceWithName(ctx, in.Spec.Name)
+	if err != nil {
+		return nil, err
+	}
+	out, err := w.reconcileNamespace(ctx, in.Spec, namespace)
+	if err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+// Reconcile a namespace, create the namespace if one does not exist, or update the namespace if one does exist.
+func (w *workflows) ReconcileNamespaces(ctx workflow.Context, in *ReconcileNamespacesInput) (*ReconcileNamespacesOutput, error) {
+	if err := validator.ValidateStruct(in); err != nil {
+		return nil, fmt.Errorf("invalid input: %s", err)
+	}
+	var (
+		getNamespacesReq = &cloudservice.GetNamespacesRequest{}
+		namespaces       = make(map[string]*namespace.Namespace)
+		out              = &ReconcileNamespacesOutput{}
+	)
+	for {
+		resp, err := w.GetNamespaces(ctx, getNamespacesReq)
+		if err != nil {
+			return nil, err
+		}
+		for i := range resp.Namespaces {
+			namespaces[resp.Namespaces[i].Spec.Name] = resp.Namespaces[i]
+		}
+		if resp.NextPageToken == "" {
+			break
+		}
+		getNamespacesReq.PageToken = resp.NextPageToken
+	}
+	for i := range in.Specs {
+		var namespace *namespace.Namespace
+		if ns, ok := namespaces[in.Specs[i].Name]; ok {
+			namespace = ns
+		}
+		// reconcile the namespace
+		o, _ := w.reconcileNamespace(ctx, in.Specs[i], namespace)
+		out.Results = append(out.Results, o)
+		// remove the reconciled namespaces from the map
+		delete(namespaces, in.Specs[i].Name)
+	}
+	// whats left in maps is only the unaccounted namespaces
```

Comment:
consider having this as optional behavior that is configurable so that someone doesn't accidentally delete all their namespaces

Reply:
nevermind, thats what the DeleteUnaccounted flag is for 🤦